File: src/main/java/at/ac/uibk/dps/cirrina/execution/command/ActionEvalCommand.java
package at.ac.uibk.dps.cirrina.execution.command;

import static at.ac.uibk.dps.cirrina.tracing.SemanticConvention.GAUGE_ACTION_EVAL_LATENCY;

import at.ac.uibk.dps.cirrina.execution.object.action.EvalAction;
import at.ac.uibk.dps.cirrina.utils.Time;
import com.google.common.flogger.FluentLogger;
import java.util.ArrayList;
import java.util.List;

public final class ActionEvalCommand extends ActionCommand {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private final EvalAction evalAction;

  ActionEvalCommand(ExecutionContext executionContext, EvalAction evalAction) {
    super(executionContext);
    this.evalAction = evalAction;
  }

  @Override
  public List<ActionCommand> execute() throws UnsupportedOperationException {
    final var start = Time.timeInMillisecondsSinceStart();

    final var commands = new ArrayList<ActionCommand>();

    try {
      final var expression = evalAction.getExpression();

      final var extent = executionContext.scope().getExtent();

      expression.execute(extent);

      // Measure latency
      final var now = Time.timeInMillisecondsSinceStart();
      final var delta = now - start;

      final var gauges = executionContext.gauges();

      gauges.getGauge(GAUGE_ACTION_EVAL_LATENCY).set(delta);
    } catch (UnsupportedOperationException e) {
      logger.atWarning().log("Expression evaluation failed");
    }

    return commands;
  }
}

File: src/main/java/at/ac/uibk/dps/cirrina/execution/command/ActionEvalCommand.kt
package at.ac.uibk.dps.cirrina.execution.command

import at.ac.uibk.dps.cirrina.execution.`object`.action.EvalAction
import io.micrometer.core.instrument.MeterRegistry

/**
 * A command responsible for evaluating a specific [EvalAction] within the provided
 * [executionContext].
 *
 * This command extracts the current scope extent and executes the underlying expression. Eval
 * actions do not generate further commands.
 *
 * @property evalAction the specific action definition containing the expression to be evaluated.
 * @property executionContext the context in which the evaluation occurs.
 * @property meterRegistry the registry used for collecting metrics.
 */
class ActionEvalCommand
internal constructor(
  private val evalAction: EvalAction,
  executionContext: ExecutionContext,
  meterRegistry: MeterRegistry,
) : ActionCommand(executionContext, meterRegistry) {

  /**
   * Executes the evaluation logic.
   *
   * @return an empty list of [ActionCommand]s.
   * @throws Exception if the command execution fails due to an internal error.
   */
  override fun execute(): List<ActionCommand> =
    evalAction.expression.execute(executionContext.scope.extent).run { emptyList() }
}
